server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoverySourceService.java

```diff
@@ -118,9 +134,14 @@ final int numberOfOngoingRecoveries() {
     }
 
     final class OngoingRecoveries {
+
         private final Map<IndexShard, ShardRecoveryContext> ongoingRecoveries = new HashMap<>();
 
+        @Nullable
+        private List<ActionListener<Void>> emptyListeners;
+
         synchronized RecoverySourceHandler addNewRecovery(StartRecoveryRequest request, IndexShard shard) {
+            assert lifecycle.started();
```

Comment:
:+1:

Reply:
When such an assertion fails, we don't have context about why the assertion failed (only that `lifecycle.started()` is not true). When we write assertions like this, it is nice to provide more context, for example: `assert lifecycle.started() : lifecycle.state()` which would display the `toString` value of `lifecycle.state()` when the assertion trips, giving us more context about the failure.